### ISSUE TITLE
Machine verification error after PostRA list scheduler

### ISSUE THREAD

**Issue by nikic** [llvm/llvm-project#52817]
The test case at https://gist.github.com/nikic/0b5ff77ab7248466eb525d934e98bbeb when run through `llc -verify-machineinstrs` results in a verifier error:

```
*** Bad machine code: Using an undefined physical register ***
- function:    wombat
- basic block: %bb.1 bb3 (0x749dca8)
- instruction: BUNDLE implicit-def dead $itstate, implicit-def $r6, implicit killed $cpsr, implicit $r0, implicit killed $r5, implicit $r8, debug-location !341; compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:0 @[ compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:204:23 @[ compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:251:5 @[ compiler-rt/lib/sanitizer_common/sanitizer_bvgraph.h:52:8 ] ] ]
- operand 4:   implicit killed $r5
```

From what I can tell, the issue is related to kill flag fixup in post-RA list scheduling done in https://github.com/llvm/llvm-project/blob/c2f2bb066b83551db3f9c32d04d6b80ac9a49a6d/llvm/lib/CodeGen/ScheduleDAGInstrs.cpp#L1139-L1153. For the bundle itself, toggleKills is called with addToLiveRegs=false, and the only use in the bundle is a debug instruction which is skipped, so the end result is that the register is not marked as live and the bundle ends up using a killed register.

**Comment by nikic** [comment]
@MatzeB @arsenm As the last two people to touch this code, maybe you have some idea of how this is supposed to work.

I'm not familiar with machine passes and unsure what the invariants in this area are. I could see a few different ways in which this could be wrong:
1. Only the kill recomputation is broken, and we should be marking the register as live for a debug use in the bundle, as the register is part of the bundle header.
2. We should not be including debug uses in the bundle header at all, as they don't seem to participate in kill tracking (?)
3. Maybe the bundle has been incorrectly constructed in the first place? The DBG_VALUE_LIST doesn't really look related to the rest of the bundle.
4. Maybe the scheduling broke something? Is there some kind of invariant that DBG_VALUE should appear prior to the last real use of the register? I see that there is code that reinserts debug intrinsics during scheduling, but I don't think it accounts for debug intrinsics in bundles.

But I don't know which one these it is (or something else entirely).

For reference, this is a larger snippet of the MIR after post-RA scheduling. Note that the `t2ADDrs` with the `$r5` use has been scheduled above the BUNDLE with the DBG_VALUE_LIST for `$r5`.
```
  renamable $r5 = t2ADDri renamable $r3, 1, 14, $noreg, $noreg
  t2STRi12 renamable $lr, renamable $r2, 0, 14, $noreg :: (store (s32) into %ir.arg2)
  t2CMPri renamable $lr, 0, 14, $noreg, implicit-def $cpsr
  renamable $r4 = t2ADDrs renamable $r0, renamable $r5, 18, 14, $noreg, $noreg
  renamable $r8 = t2ADDrr killed renamable $r4, renamable $r12, 14, $noreg, $noreg
  renamable $r4 = nuw t2LSLrr renamable $r9, renamable $r3, 14, $noreg, $noreg
  renamable $r3 = t2ANDrr killed renamable $r3, killed renamable $r4, 14, $noreg, $noreg
  renamable $r4 = t2ADDrs renamable $r1, killed renamable $r5, 18, 14, $noreg, $noreg
  BUNDLE implicit-def dead $itstate, implicit-def $r6, implicit killed $cpsr, implicit $r0, implicit killed $r5, implicit $r8, debug-location !26; compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:0 @[ compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:204:23 @[ compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:251:5 @[ compiler-rt/lib/sanitizer_common/sanitizer_bvgraph.h:52:8 ] ] ] {
    t2IT 1, 6, implicit-def $itstate
    renamable $r6 = t2MOVi 0, 1, $cpsr, $noreg, implicit internal $itstate
    DBG_VALUE_LIST !"this", !DIExpression(DW_OP_LLVM_arg, 0, DW_OP_plus_uconst, 135168, DW_OP_LLVM_arg, 1, DW_OP_constu, 4, DW_OP_mul, DW_OP_plus, DW_OP_plus_uconst, 4, DW_OP_stack_value), $r0, $r5, debug-location !26; compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:0 @[ compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:204:23 @[ compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:251:5 @[ compiler-rt/lib/sanitizer_common/sanitizer_bvgraph.h:52:8 ] ] ] line no:0
    t2STRi12 internal renamable $r6, renamable $r8, 0, 1, $cpsr, implicit internal $itstate, debug-location !36 :: (store (s32) into %ir.tmp10); compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:28:24 @[ compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:204:23 @[ compiler-rt/lib/sanitizer_common/sanitizer_bitvector.h:251:5 @[ compiler-rt/lib/sanitizer_common/sanitizer_bvgraph.h:52:8 ] ] ]
    renamable $r6 = t2MOVi 1, 0, killed $cpsr, $noreg, implicit internal killed $r6, implicit internal killed $itstate
  }
```

**Comment by MatzeB** [comment]
Take this with a grain of salt, as I am not actively using bundles (and I remember struggling to capture their exact expected semantics in the past).

AFAIK the bundle header should reproduce the effects on register state after executing the bundle. DBG_VALUE uses do not have effects, they are not consider a read of the register and if its not a read then it certainly cannot be a kill either.

I guess we could have a discussion whether the bundle header is allowed to state more effects than are actually happening within the bundle. But unless someone has a convincing use-case I would aim for having the bundle header exactly reproduce the effects, from that interpretation the bundle was not formed properly in your example. (Isn't `DBG_VALUE_LIST` a relatively new addition in CodeGen? Maybe if-conversion gets something wrong there...)

I guess in this ARM case where the bundle still gets expanded into multiple instrunctions in machine code, we may need to allow debug instruction inside the bundle between the instructions...

**Comment by bjope** [comment]
FWIW, in our downstream target we avoid putting DBG_VALUE etc inside bundles (the downstream scheduler/packetizer will make sure the DBG-instructions are moved outside the bundle). I guess that is why we've never stumbled upon this problem.

Our design is based on that the debugger never will stop in the middle of a bundle anyway (execution can only step one whole bundle at a time). So not much point in having DBG-instructions inside the bundles. Well, a possible drawback is that we can't really model that some value was put in a register just before a call in case the call is bundled with something that sets up a variables value in a register. But how to encode that in DWARF given that the DBG_VALUE* is inside the bundle might be complicated as well (I can't remember that the dwarf debug emitter handles that nicely anyway).

With that being said, it seems like other targets doesn't prevent the DBG-instructions from ending up inside a bundle so I do not object to the proposed fix in D116945. Just trying to share some info about how we deal with BUNDLE vs DBG in our target.
